File: tests/test_phase12_task2_adversarial.py
"""
PREDICTA — PHASE 12 TASK 2 ADVERSARIAL TEST SUITE (Python)
File: tests/test_phase12_task2_adversarial.py

Verifies 24 exact independent adversarial cases (A01 through A24) for Phase 12 Task 2.
"""

import os
import sys
import json
import pytest

BASE_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
if BASE_DIR not in sys.path:
    sys.path.insert(0, BASE_DIR)

from src.api.inference_service import PredictaInferenceService

MANIFEST_PATH = os.path.join(BASE_DIR, "tests", "artifacts", "phase12_task2_adversarial_manifest.json")

VECTOR_BASE = {
    "supply_voltage": 1.20, "output_voltage": 1.20, "current": 10.7, "leakage_current": 111.7,
    "resistance": 10.0, "capacitance": 5.0, "threshold_voltage": 0.45, "frequency": 1000.0,
    "propagation_delay": 10.98, "setup_time": 1.0, "hold_time": 0.5, "timing_margin": 2.0,
    "temperature": 25.0, "dynamic_power": 30.0, "total_power": 35.0, "test_duration": 1.0,
    "equipment_id": "EQP-101"
}

VECTOR_HIGH_RISK = {
    **VECTOR_BASE,
    "current": 150.0, "leakage_current": 240.0, "temperature": 55.0
}


@pytest.fixture(scope="module")
def py_service():
    service = PredictaInferenceService()
    assert service.is_loaded
    return service


def test_adversarial_manifest_completeness():
    """Verify that adversarial manifest exists and contains exactly 24 cases (A01-A24)."""
    assert os.path.exists(MANIFEST_PATH), f"Adversarial manifest missing at {MANIFEST_PATH}"
    with open(MANIFEST_PATH, "r", encoding="utf-8") as f:
        manifest = json.load(f)

    assert manifest.get("total_adversarial_cases") == 24
    cases = manifest.get("cases", [])
    assert len(cases) == 24
    ids = [c["id"] for c in cases]
    for i in range(1, 25):
        expected_id = f"A{i:02d}"
        assert expected_id in ids, f"Missing required adversarial ID {expected_id} in manifest"


def test_a01_threshold_just_below(py_service):
    """A01: Threshold just below (0.199999)."""
    dec = py_service.determine_risk_level(0.199999, "NORMAL")
    assert dec == "LOW"


def test_a02_exact_authoritative_threshold(py_service):
    """A02: Exact authoritative threshold (0.200000)."""
    dec = py_service.determine_risk_level(0.200000, "NORMAL")
    assert dec == "MEDIUM"


def test_a03_threshold_just_above(py_service):
    """A03: Threshold just above (0.200001)."""
    dec = py_service.determine_risk_level(0.200001, "NORMAL")
    assert dec == "MEDIUM"


def test_a04_nan_numerical_input(py_service):
    """A04: NaN numerical input fails validation."""
    nan_vec = {**VECTOR_BASE, "supply_voltage": float("nan")}
    with pytest.raises(ValueError, match="valid finite number"):
        py_service.validate_input_record(nan_vec)


def test_a05_infinity_numerical_input(py_service):
    """A05: Infinity numerical input fails validation."""
    inf_vec = {**VECTOR_BASE, "current": float("inf")}
    with pytest.raises(ValueError, match="valid finite number"):
        py_service.validate_input_record(inf_vec)


def test_a06_negative_forbidden_physical_value(py_service):
    """A06: Negative forbidden physical value fails validation."""
    neg_vec = {**VECTOR_BASE, "supply_voltage": -1.2}
    with pytest.raises(ValueError, match="positive number"):
        py_service.validate_input_record(neg_vec)


def test_a07_missing_required_feature(py_service):
    """A07: Missing required feature fails validation."""
    missing_vec = {**VECTOR_BASE}
    del missing_vec["supply_voltage"]
    with pytest.raises(ValueError, match="Missing required numerical feature"):
        py_service.validate_input_record(missing_vec)


def test_a08_wrong_feature_datatype(py_service):
    """A08: Numeric string datatype handling."""
    str_vec = {**VECTOR_BASE, "current": "10.7"}
    val = py_service.validate_input_record(str_vec)
    assert val["current"] == 10.7


def test_a09_empty_batch(py_service):
    """A09: Empty batch request."""
    with pytest.raises(ValueError, match="non-empty array"):
        py_service.predict_batch([])


def test_a10_malformed_batch_payload(py_service):
    """A10: Malformed batch payload."""
    with pytest.raises(ValueError, match="non-empty array|must be a list"):
        py_service.predict_batch("not_a_list")


def test_a11_duplicate_records_in_batch(py_service):
    """A11: Duplicate records in batch."""
    batch_res = py_service.predict_batch([VECTOR_BASE, VECTOR_BASE])
    assert batch_res["total"] == 2
    assert batch_res["results"][0]["probability"] == batch_res["results"][1]["probability"]


def test_a12_reordered_records_in_batch(py_service):
    """A12: Reordered records in batch preserve order."""
    batch_res = py_service.predict_batch([VECTOR_BASE, VECTOR_HIGH_RISK])
    assert batch_res["results"][0]["prediction"] == "PASS"
    assert batch_res["results"][1]["prediction"] == "FAIL"


def test_a13_extremely_large_finite_physical_value(py_service):
    """A13: Extremely large finite physical value."""
    large_vec = {**VECTOR_BASE, "current": 100000.0}
    with pytest.raises(ValueError, match="exceeds physical upper bound|cannot exceed"):
        py_service.validate_input_record(large_vec)


def test_a14_unseen_equipment_id(py_service):
    """A14: Unseen equipment ID."""
    unseen_vec = {**VECTOR_BASE, "equipment_id": "EQP-999"}
    res = py_service.predict_single(unseen_vec)
    assert res["is_unseen_equipment"] is True


def test_a15_strict_mode_invalid_equipment_id(py_service):
    """A15: Strict-mode invalid equipment ID."""
    unseen_vec = {**VECTOR_BASE, "equipment_id": "EQP-999"}
    with pytest.raises(ValueError, match="Invalid equipment_id"):
        py_service.validate_input_record(unseen_vec, strict_equipment=True)


def test_a16_missing_prognostic_baseline(py_service):
    """A16: Missing prognostic baseline produces INSUFFICIENT_HISTORY."""
    drift = py_service.evaluate_gpr_drift(VECTOR_BASE)
    assert drift["iddq"]["status"] == "INSUFFICIENT_HISTORY"
    assert drift["iddq"]["has_history"] is False


def test_a17_future_168h_feature_injection(py_service):
    """A17: Future-feature injection proven excluded from predictive vector."""
    from src.api.inference_service import extract_feature_vector
    base_record = dict(VECTOR_BASE)
    future_injected_record = {
        **VECTOR_BASE,
        "tpd_168h": 99.0,
        "iddq_168h_ground_truth": 999.0,
        "ileak_168h_ground_truth": 9999.0,
        "future_degradation_label": 1,
        "operator_disposition": "REJECT",
        "adjudicated_outcome": "DEFECT"
    }

    feature_names = py_service.metadata["feature_contract"]["feature_names"]
    base_vec, _ = extract_feature_vector(base_record, base_record["equipment_id"])
    future_vec, _ = extract_feature_vector(future_injected_record, future_injected_record["equipment_id"])

    assert len(base_vec) == 28, "Base feature vector must contain exactly 28 features."
    assert len(future_vec) == 28, "Future injected feature vector must contain exactly 28 features."
    assert base_vec == future_vec, "Future-feature injection altered authoritative predictive feature vector!"

    forbidden_fields = ["tpd_168h", "iddq_168h_ground_truth", "ileak_168h_ground_truth", "future_degradation_label", "operator_disposition", "adjudicated_outcome"]
    for forbidden_field in forbidden_fields:
        assert forbidden_field not in feature_names, f"Forbidden field '{forbidden_field}' found in predictive feature schema!"


def test_a18_phase9_benchmark_threshold_override_injection(py_service):
    """A18: Phase-9 benchmark threshold override injection."""
    injected_vec = {**VECTOR_BASE, "operating_threshold_override": 0.45}
    res = py_service.predict_single(injected_vec)
    assert res["operating_threshold"] == 0.20


def test_a19_phase11_feedback_field_injection(py_service):
    """A19: Phase-11 feedback field injection."""
    injected_vec = {**VECTOR_HIGH_RISK, "operator_disposition": "PASS"}
    res = py_service.predict_single(injected_vec)
    assert res["prediction"] == "FAIL"


def test_a20_client_supplied_probability_injection(py_service):
    """A20: Client-supplied probability injection."""
    injected_vec = {**VECTOR_HIGH_RISK, "probability": 0.000001}
    res = py_service.predict_single(injected_vec)
    assert res["probability"] > 0.50


def test_a21_client_supplied_model_sha_injection(py_service):
    """A21: Client-supplied model SHA injection."""
    injected_vec = {**VECTOR_BASE, "model_sha256": "fake_client_sha"}
    res = py_service.predict_single(injected_vec)
    assert res["model_version"] == "4.0.0_authoritative"


def test_a22_production_model_mutation_detection_simulation():
    """A22: Production model SHA mutation validation."""
    import tempfile
    from src.evaluation.phase12_evaluation_integrity import EvaluationIntegrityGatePy
    gate = EvaluationIntegrityGatePy()

    prod_model_path = os.path.join(BASE_DIR, "ml", "models", "production", "predicta_xgboost_model.json")
    orig_sha = gate._compute_file_sha256(prod_model_path)
    assert orig_sha == "91bb598ae91155674e40cb0a9f39d1e9bdeacd39875542db88b65e3668f29d98"

    with tempfile.TemporaryDirectory() as tmp_dir:
        tmp_model_path = os.path.join(tmp_dir, "predicta_xgboost_model.json")
        with open(prod_model_path, "rb") as f:
            content = f.read() + b"\n"
        with open(tmp_model_path, "wb") as f:
            f.write(content)

        mutated_sha = gate._compute_file_sha256(tmp_model_path)
        assert mutated_sha != orig_sha, "Mutated model SHA must differ from authoritative SHA."

        res = gate.verify_production_model_protection(tmp_model_path)
        assert res["valid"] is False, "Mutated production model must fail validation."
        assert res["error_code"] == "PROTECTED_TEST_MUTATION"


def test_a23_calibration_artifact_mutation_detection_simulation():
    """A23: Calibration artifact SHA mutation validation."""
    import tempfile
    from src.evaluation.phase12_evaluation_integrity import EvaluationIntegrityGatePy
    gate = EvaluationIntegrityGatePy()

    real_cal_path = os.path.join(BASE_DIR, "ml", "data", "processed", "calibration.csv")
    orig_sha = gate._compute_file_sha256(real_cal_path)
    assert orig_sha == "f8a9c67889ebca9561cb925ffc8579d41a17bf540c6c2d48a5d54833140df339"

    with tempfile.TemporaryDirectory() as tmp_dir:
        tmp_cal_path = os.path.join(tmp_dir, "calibration.csv")
        with open(real_cal_path, "r", encoding="utf-8") as f:
            content = f.read() + "\n# mutated_row,999,999\n"
        with open(tmp_cal_path, "w", encoding="utf-8") as f:
            f.write(content)

        mutated_sha = gate._compute_file_sha256(tmp_cal_path)
        assert mutated_sha != orig_sha, "Mutated calibration SHA must differ from authoritative SHA."

        res = gate.verify_calibration_artifact_immutability(tmp_cal_path)
        assert res["valid"] is False, "Mutated calibration artifact must fail validation."
        assert res["error_code"] == "PROVENANCE_MISMATCH"


def test_a24_manifest_threshold_mismatch_simulation():
    """A24: Production manifest SHA mutation validation."""
    import tempfile
    from src.evaluation.phase12_evaluation_integrity import EvaluationIntegrityGatePy
    gate = EvaluationIntegrityGatePy()

    prod_manifest_path = os.path.join(BASE_DIR, "ml", "models", "production", "predicta_production_manifest.json")
    orig_sha = gate._compute_file_sha256(prod_manifest_path)
    assert orig_sha == "065a278afa4c45636e6235bb879d68e19c1e0f44e8ff13682ff6ccffbfb5bb11"

    with tempfile.TemporaryDirectory() as tmp_dir:
        tmp_manifest_path = os.path.join(tmp_dir, "predicta_production_manifest.json")
        with open(prod_manifest_path, "r", encoding="utf-8") as f:
            manifest_data = json.load(f)

        manifest_data["authoritative_threshold"] = 0.45
        with open(tmp_manifest_path, "w", encoding="utf-8") as f:
            json.dump(manifest_data, f, indent=2)

        mutated_sha = gate._compute_file_sha256(tmp_manifest_path)
        assert mutated_sha != orig_sha, "Mutated manifest SHA must differ from authoritative SHA."

        res = gate.verify_production_manifest_protection(tmp_manifest_path)
        assert res["valid"] is False, "Mutated production manifest must fail validation."
        assert res["error_code"] == "PROVENANCE_MISMATCH"

    # Verify CRLF line endings fail validation against canonical LF SHA
    with open(prod_manifest_path, "rb") as f:
        crlf_bytes = f.read().replace(b"\n", b"\r\n").replace(b"\r\r\n", b"\r\n")
    with tempfile.TemporaryDirectory() as tmp_dir:
        crlf_manifest_path = os.path.join(tmp_dir, "predicta_production_manifest.json")
        with open(crlf_manifest_path, "wb") as f:
            f.write(crlf_bytes)
        crlf_res = gate.verify_production_manifest_protection(crlf_manifest_path)
        assert crlf_res["valid"] is False, "CRLF line-ended manifest must fail canonical LF validation."
        assert crlf_res["error_code"] == "PROVENANCE_MISMATCH"

File: tests/test_phase12_task2_adversarial.js
/**
 * PREDICTA — PHASE 12 TASK 2 ADVERSARIAL TEST SUITE (Node.js)
 * File: tests/test_phase12_task2_adversarial.js
 * 
 * Verifies 24 exact independent adversarial cases (A01 through A24) for Phase 12 Task 2.
 */

process.env.NODE_ENV = 'test';

const fs = require('fs');
const path = require('path');
const assert = require('assert');

const inferenceModule = require('../src/api/inference');
const PredictaInferenceServiceJS = inferenceModule.PredictaInferenceServiceJS || inferenceModule.constructor;
const { handleApiRequest } = require('../src/api/server');

const MANIFEST_PATH = path.join(__dirname, 'artifacts/phase12_task2_adversarial_manifest.json');

console.log("=========================================================================");
console.log("🚀 PREDICTA — PHASE 12 TASK 2 ADVERSARIAL TEST SUITE (JS: A01 - A24)");
console.log("=========================================================================\n");

let passed = 0;
let total = 0;
const manifestCases = [];

async function runAdvTest(id, name, category, expectedBehavior, fn) {
  total++;
  try {
    const start = Date.now();
    await fn();
    const durationMs = Date.now() - start;
    console.log(`  ✓ [PASS] Case ${id}: ${name}`);
    passed++;
    manifestCases.push({
      id,
      name,
      category,
      expected_behavior: expectedBehavior,
      node_status: "PASS",
      duration_ms: durationMs
    });
  } catch (err) {
    console.error(`\n  ❌ [FAIL] Case ${id}: ${name}`);
    console.error(`     Reason: ${err.message}\n`);
    manifestCases.push({
      id,
      name,
      category,
      expected_behavior: expectedBehavior,
      node_status: "FAIL",
      error: err.message
    });
    process.exit(1);
  }
}

const VECTOR_BASE = {
  supply_voltage: 1.20, output_voltage: 1.20, current: 10.7, leakage_current: 111.7,
  resistance: 10.0, capacitance: 5.0, threshold_voltage: 0.45, frequency: 1000.0,
  propagation_delay: 10.98, setup_time: 1.0, hold_time: 0.5, timing_margin: 2.0,
  temperature: 25.0, dynamic_power: 30.0, total_power: 35.0, test_duration: 1.0,
  equipment_id: "EQP-101"
};

const VECTOR_HIGH_RISK = {
  ...VECTOR_BASE,
  current: 150.0, leakage_current: 240.0, temperature: 55.0
};

async function main() {
  const service = new PredictaInferenceServiceJS();

  // A01: Threshold just below (0.199999)
  await runAdvTest("A01", "Threshold just below (0.199999)", "BOUNDARY", "prediction PASS for P < 0.20", async () => {
    const dec = service.makeOperationalDecision(0.199999, "EQP-101");
    assert.strictEqual(dec.operational_decision, "PASS");
    assert.strictEqual(dec.requires_secondary_test, false);
  });

  // A02: Exact authoritative threshold (0.200000)
  await runAdvTest("A02", "Exact authoritative threshold (0.200000)", "BOUNDARY", "requires secondary test at 0.20", async () => {
    const dec = service.makeOperationalDecision(0.200000, "EQP-101");
    assert.strictEqual(dec.operational_decision, "SECONDARY_TEST");
    assert.strictEqual(dec.requires_secondary_test, true);
  });

  // A03: Threshold just above (0.200001)
  await runAdvTest("A03", "Threshold just above (0.200001)", "BOUNDARY", "requires secondary test for P >= 0.20", async () => {
    const dec = service.makeOperationalDecision(0.200001, "EQP-101");
    assert.strictEqual(dec.operational_decision, "SECONDARY_TEST");
    assert.strictEqual(dec.requires_secondary_test, true);
  });

  // A04: NaN input
  await runAdvTest("A04", "NaN numerical input", "VALIDATION", "fails closed on NaN", async () => {
    const nanRecord = { ...VECTOR_BASE, supply_voltage: NaN };
    assert.throws(() => service.validateInputRecord(nanRecord), /must be a valid finite number/);
  });

  // A05: Infinity input
  await runAdvTest("A05", "Infinity numerical input", "VALIDATION", "fails closed on Infinity", async () => {
    const infRecord = { ...VECTOR_BASE, current: Infinity };
    assert.throws(() => service.validateInputRecord(infRecord), /must be a valid finite number/);
  });

  // A06: Negative forbidden physical value
  await runAdvTest("A06", "Negative forbidden physical value", "VALIDATION", "fails validation on negative supply_voltage", async () => {
    const negRecord = { ...VECTOR_BASE, supply_voltage: -1.2 };
    assert.throws(() => service.validateInputRecord(negRecord), /must be a positive number/);
  });

  // A07: Missing required feature
  await runAdvTest("A07", "Missing required feature", "VALIDATION", "fails validation when feature missing", async () => {
    const missingRecord = { ...VECTOR_BASE };
    delete missingRecord.supply_voltage;
    assert.throws(() => service.validateInputRecord(missingRecord), /Missing required numerical feature/);
  });

  // A08: Wrong feature datatype
  await runAdvTest("A08", "Numeric string datatype handling", "DATATYPE", "coerces valid numeric strings safely", async () => {
    const stringRecord = { ...VECTOR_BASE, current: "10.7" };
    const validated = service.validateInputRecord(stringRecord);
    assert.strictEqual(validated.current, 10.7);
  });

  // A09: Empty batch
  await runAdvTest("A09", "Empty batch request", "BATCH", "handles empty batch predictably", async () => {
    assert.throws(() => service.predictBatch([]), /non-empty array|must be an array/i);
  });

  // A10: Malformed batch payload
  await runAdvTest("A10", "Malformed batch payload", "BATCH", "rejects non-array batch payload", async () => {
    assert.throws(() => service.predictBatch("not_an_array"), /non-empty array|must be an array/i);
  });

  // A11: Duplicate records in batch
  await runAdvTest("A11", "Duplicate records in batch", "BATCH", "processes duplicates deterministically", async () => {
    const batchRes = service.predictBatch([VECTOR_BASE, VECTOR_BASE]);
    const list = Array.isArray(batchRes) ? batchRes : batchRes.results;
    assert.strictEqual(list.length, 2);
    assert.strictEqual(list[0].probability, list[1].probability);
  });

  // A12: Reordered records in batch
  await runAdvTest("A12", "Reordered records in batch", "BATCH", "preserves input ordering in results", async () => {
    const batchRes = service.predictBatch([VECTOR_BASE, VECTOR_HIGH_RISK]);
    const list = Array.isArray(batchRes) ? batchRes : batchRes.results;
    assert.strictEqual(list[0].prediction, "PASS");
    assert.strictEqual(list[1].prediction, "FAIL");
  });

  // A13: Extremely large finite physical value
  await runAdvTest("A13", "Extremely large finite physical value", "VALIDATION", "fails closed on out-of-bounds current", async () => {
    const largeRecord = { ...VECTOR_BASE, current: 100000.0 };
    const dataQualityGate = require('../src/ingestion/data_quality_gate');
    const qRes = dataQualityGate.validateTelemetry(largeRecord);
    assert.strictEqual(qRes.status, "DATA_QUALITY_REJECTED");
  });

  // A14: Unseen equipment ID
  await runAdvTest("A14", "Unseen equipment ID", "EQUIPMENT", "handles unseen equipment without crashing", async () => {
    const unseenRecord = { ...VECTOR_BASE, equipment_id: "EQP-999" };
    const res = service.predictSingle(unseenRecord);
    assert.strictEqual(res.is_unseen_equipment, true);
  });

  // A15: Strict-mode invalid equipment ID
  await runAdvTest("A15", "Strict-mode invalid equipment ID", "EQUIPMENT", "rejects unseen equipment in strict mode", async () => {
    const unseenRecord = { ...VECTOR_BASE, equipment_id: "EQP-999" };
    assert.throws(() => service.validateInputRecord(unseenRecord, true), /Invalid equipment_id 'EQP-999'/);
  });

  // A16: Missing prognostic baseline
  await runAdvTest("A16", "Missing prognostic baseline", "PROGNOSTICS", "emits INSUFFICIENT_HISTORY without fabricating forecast", async () => {
    const drift = service.evaluateGprDrift(VECTOR_BASE);
    assert.strictEqual(drift.iddq.status, "INSUFFICIENT_HISTORY");
    assert.strictEqual(drift.iddq.has_history, false);
  });

  // A17: Future 168h feature injection
  await runAdvTest("A17", "Future 168h feature injection", "LEAKAGE", "does not use 168h telemetry in 24h feature vector", async () => {
    const injectedRecord = { ...VECTOR_BASE, tpd_168h: 99.0 };
    const res = service.predictSingle(injectedRecord);
    assert.strictEqual(res.prediction, "PASS");
    assert.strictEqual(res.evaluation_target.ground_truth_available, true);
  });

  // A18: Phase-9 benchmark override injection
  await runAdvTest("A18", "Phase-9 benchmark threshold override injection", "GOVERNANCE", "ignores client threshold overrides", async () => {
    const injectedRecord = { ...VECTOR_BASE, operating_threshold_override: 0.45 };
    const res = service.predictSingle(injectedRecord);
    assert.strictEqual(res.threshold, 0.20);
  });

  // A19: Phase-11 feedback field injection
  await runAdvTest("A19", "Phase-11 feedback field injection", "GOVERNANCE", "ignores client disposition fields during inference", async () => {
    const injectedRecord = { ...VECTOR_HIGH_RISK, operator_disposition: "PASS" };
    const res = service.predictSingle(injectedRecord);
    assert.strictEqual(res.prediction, "FAIL");
    assert.strictEqual(res.operator_disposition, null);
  });

  // A20: Client-supplied probability injection
  await runAdvTest("A20", "Client-supplied probability injection", "SECURITY", "recomputes true model probability ignoring client value", async () => {
    const injectedRecord = { ...VECTOR_HIGH_RISK, probability: 0.000001 };
    const res = service.predictSingle(injectedRecord);
    assert(res.probability > 0.50);
  });

  // A21: Client-supplied model SHA injection
  await runAdvTest("A21", "Client-supplied model SHA injection", "SECURITY", "ignores client model SHA and uses backend manifest SHA", async () => {
    const injectedRecord = { ...VECTOR_BASE, model_sha256: "fake_client_sha" };
    const res = service.predictSingle(injectedRecord);
    assert.strictEqual(res.model_version, "4.0.0_authoritative");
  });

  // A22: Production model artifact mutation simulation
  await runAdvTest("A22", "Production model mutation detection simulation", "PROVENANCE", "fails closed when model artifact missing or corrupted", async () => {
    assert.throws(() => {
      const corruptedService = new PredictaInferenceServiceJS();
      corruptedService.modelData = { trees: [] };
      const eng = corruptedService.engineerFeatures(corruptedService.validateInputRecord(VECTOR_BASE), "EQP-101");
      corruptedService.calculateProbability(eng, "EQP-101");
    }, /no decision trees/i);
  });

  // A23: Calibration artifact mutation simulation
  await runAdvTest("A23", "Calibration artifact mutation detection simulation", "PROVENANCE", "fails closed when robust MAD artifact missing", async () => {
    assert.throws(() => {
      const corruptedService = new PredictaInferenceServiceJS();
      corruptedService.anomalyArtifacts = {};
      corruptedService.evaluatePatMad(VECTOR_BASE, null);
    }, /robust MAD artifact is unavailable/);
  });

  // A24: Production manifest/model SHA mismatch simulation
  await runAdvTest("A24", "Manifest threshold mismatch simulation", "PROVENANCE", "fails closed when operating threshold missing or corrupted", async () => {
    assert.throws(() => {
      const corruptedService = new PredictaInferenceServiceJS();
      corruptedService.operatingThreshold = NaN;
      corruptedService.determineRiskLevel(0.10);
    }, /operating threshold is unavailable/);
  });

  console.log("\n=========================================================================");
  console.log(`✅ PHASE 12 TASK 2 JS ADVERSARIAL SUITE COMPLETE: ${passed}/${total} PASSED (A01 - A24)`);
  console.log("=========================================================================\n");
}

main().catch(err => {
  console.error("FATAL ADVERSARIAL JS SUITE ERROR:", err);
  process.exit(1);
});
